Support for MS8607 PHT (Pressure Humidity Temperature) sensor

e28eta opened the pull request:
# What does this implement/fix?

The MS8607 ([datasheet](https://www.te.com/commerce/DocumentDelivery/DDEController?Action=showdoc&DocId=Data+Sheet%7FMS8607-02BA01%7FB3%7Fpdf%7FEnglish%7FENG_DS_MS8607-02BA01_B3.pdf%7FCAT-BLPS0018)) is a combined Pressure, Humidity, and Temperature sensor. It communicates over i2c, and has two different addresses (one for PT & the other for H).

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes esphome/feature-requests#932

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#1974

## Test Environment

- [x] ESP32
- [ ] ESP32 IDF
- [ ] ESP8266

I don't know why it wouldn't work on all of these, but I don't have an ESP8266, and I haven't done anything with ESP32 IDF.

## Example entry for `config.yaml`:

```yaml
# Example config.yaml
sensor:
  - platform: ms8607
    temperature:
      name: Temperature
    humidity:
      name: Humidity
    pressure:
      name: Pressure
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

e28eta committed:
initial work toward ms8607 support

- I think sensor.py configuration & validation is good
- adapting the ms5611 c++ code for ms8607. I think it'll setup & read PROM correctly,
but update() does nothing.

Primary question right now is whether I'm reading PROM correctly with the 16 bit version

e28eta committed:
Fix compiler errors in cpp code

e28eta committed:
note to self re: introducting segfaults

e28eta committed:
Record & log failure reason during setup; log humidity address; simplify PROM read

- Keeping track of the reason why setup failed, and including that in the dump_config output
- Keep humidity address in protected variable, since I cannot read it from the I2CDevice
- Verified that, as I thought, reading 16 byte values resulted in MSB ordering.

e28eta committed:
fixup: Attempt to fix compile error

e28eta committed:
attempt to fix compiler warning on failure reason logging

e28eta committed:
Add prom CRC algorithm, and check that it's a match

e28eta committed:
Adding struct for calibration values, and storing them

e28eta committed:
fixup: add prom_ back in so it will compile

e28eta committed:
Rename FailureReason to ErrorCode

Matches ~13 other components, instead of being a unique name

e28eta committed:
Beginnings of code to read humidity

Testing out the humidity resolution constants, since datasheet & other implementations
differ.

No checksum support (not in datasheet?), and not publishing the values.

e28eta committed:
fix some compiler errors

e28eta committed:
couple more compile errors

e28eta committed:
attempt to fix percent calc that shows -6 every time

e28eta committed:
still trying to fix percent

e28eta committed:
does this matter?

e28eta committed:
Strip OSR configurability.

Didn't seem worth the extra complexity to me

e28eta committed:
attempt to remove conversion value, see if that triggers some humidity not ready msgs

e28eta committed:
First pass at pressure & temperature implementation

Interesting that some of the constants/calculations varied from the source sensor, where
it's usually been pretty close.

e28eta committed:
Add some recommended parentheses

e28eta committed:
publish humidity reading

e28eta committed:
Fix compile errors for rebase

e28eta committed:
Update sensor.py with kwargs and async to_code

e28eta committed:
Add additional error case

e28eta committed:
Add some more comments on re-reading the code to re-understand

e28eta committed:
Humidity: turn into non-blocking, compensate for temperature, check CRC

e28eta committed:
Some cleanup of logging, protect against null humidity_sensor_

e28eta committed:
Eliminate delay(15) during setup() call, retry RESET commands 3x if failed

grep didn't find _any_ components using `set_retry`, but I think this is correct.
There's a little bit of the pyramid of doom and/or out-of-order code in `setup()`
due to scheduling future code and then current code showing up below it, but I feel okay
about this right now.

e28eta committed:
Rewrite some of the log messages

e28eta committed:
mark_failed() if setup fails too many times

e28eta committed:
Add line to test5.yaml to ensure component compiles correctly

(per CONTRIBUTING.md)

e28eta committed:
Remove unused humidity register commands, add comment w/set_retry intervals

If this Component supported changing the OSR for humidity, would need to be able to
read & write the humidity register. I'm not implementing that right now, might as well
drop the dead code.

e28eta committed:
Move remaining setup attempt counter into static local

I didn't know this was possible until I saw it in https://github.com/esphome/esphome/pull/2721

I'd prefer https://github.com/e28eta/esphome/pull/1 - but I think this is a small cleanup

e28eta committed:
Update sensor.py: remove TODO that isnt, update accuracy based on datasheet resolution

e28eta committed:
Add links to datasheet and sample implementations

e28eta committed:
Adding self as codeowner per probot-esphome request

e28eta committed:
Apply clang-format suggestions from CI

e28eta committed:
clang-tidy suggestions: fixed TAG const, method name, idx type

e28eta committed:
clang-tidy: fix (?) new/delete -> make_unique ptr

I've never used this API before, but I think this is correct

e28eta commented:
Known next steps:

- [ ] pass CI checks ❤️ 
- [ ] pass code review
- [ ] create esphome/esphomedocs pull request
- [ ] verification that addition to `test5.yaml` is sufficient for "writing tests" step
- [ ] change PR if e28eta/esphome#1 is accepted

e28eta commented on esphome/components/ms8607/ms8607.cpp:
```diff
@@ -0,0 +1,430 @@
+
+void MS8607Component::set_humidity_sensor_address(uint8_t address) {
+  this->humidity_sensor_address_ = address;
+  this->humidity_i2c_device_ = make_unique<I2CDevice>();
```
I think I've correctly created the `I2CDevice`. clang-tidy yelled at me for using `new`.

This is being done in C++ instead of the generated code from `sensor.py` because I didn't want it to be possible to give them different `i2c_bus` objects. The two `I2CDevices` are literally connected to the same pins on the MS8607, but respond to different addresses.

jesserockz replied:
This looks fine.

jesserockz replied:
After coming back to this PR, I think I would prefer if the humidity sensor had its own class that extended I2CDevice and it was created from codegen using `i2c.register_i2c_device`. I don't think the end user will give the two sensors a different I2CBus, and I would rather avoid small "hacks" like this in the codebase that might cause issues later on.

e28eta commented on esphome/components/ms8607/ms8607.cpp:
```diff
@@ -0,0 +1,430 @@
+        this->error_code_ = ErrorCode::NONE;
+        this->status_clear_error();
+
+        // 15ms delay matches datasheet, Adafruit_MS8607 & SparkFun_PHT_MS8607_Arduino_Library
```
The datasheet says "on startup, send reset, wait 15ms, then you can interact".

My first pass at this just had `delay(15); //NOLINT`, but when I looked at what it'd take to use `set_timeout` + skipping `update()` until finished, it didn't seem that bad.

It also allowed `set_retry` to be woven in pretty easily. I like that because I **have** seen occasional `NACK` to the initial reset commands, and that's why I put this PR on the back burner for ~18 months: it worked, but not always, and I haven't been able to figure out why. The esphome device almost never restarts, and so it was workable for me but not "production-ready"

e28eta commented on esphome/components/ms8607/ms8607.cpp:
```diff
@@ -0,0 +1,430 @@
+static const uint8_t MS8607_CMD_ADC_READ = 0x00;
+
+// TODO: allow OSR to be turned down for speed and/or lower power consumption via configuration.
+// ms8607 supports 6 different settings
```
I don't actually plan to do this anytime (soon? ever?)... but it is possible.

e28eta commented on esphome/components/ms8607/ms8607.cpp:
```diff
@@ -0,0 +1,430 @@
+ * @brief Calculates CRC value for the provided humitity (+ status bits) value
+ *
+ * CRC-8 check comes from other MS8607 libraries on github. I did not find it in the datasheet,
+ * and it differs from the crc8 implementation that's already part of esphome.
```
I really wanted to just use the existing implementation of `crc8`, especially since I think they're both using [CRC-8-Dallas/Maxim](https://en.wikipedia.org/wiki/Cyclic_redundancy_check#Polynomial_representations_of_cyclic_redundancy_checks) in different forms ("reversed" vs "reversed reciprocal" based on the polynomials on wikipedia).

https://github.com/esphome/esphome/blob/e621b938e36b0b6ede82d14419a4830529789659/esphome/core/helpers.cpp#L50-L64

However, I don't actually understand much of what I was reading, and couldn't figure out how to use the existing `crc8` function on the bytes returned from the humidity sensor to verify the CRC provided. 😞

e28eta commented on esphome/components/ms8607/ms8607.cpp:
```diff
@@ -0,0 +1,430 @@
+  }
+  if (!(humidity & 0x2)) {
+    // data sheet says Bit1 should always set, but nothing about what happens if it isn't
+    ESP_LOGE(TAG, "Humidity status bit was not set to 1?");
```
I had trouble deciding whether to leave this log message in, fail the reading if it doesn't match, or skip the comparison completely since a failure is essentially ignored.

But this is where I ended up

e28eta committed:
Adding better logging for error_code like rc522 does

e28eta commented on esphome/components/ms8607/sensor.py:
```diff
@@ -0,0 +1,77 @@
+    UNIT_PERCENT,
+)
+
+DEPENDENCIES = ["i2c"]
```
Any guidance on whether this should be `DEPENDENCIES` or `AUTO_LOAD`?

I see that most/all of the `I2CDevice` components use `DEPENDENCIES`, but since it works with _just_ `i2c:` in the yaml file it seems to me like this could/should be `AUTO_LOAD` to me.

jesserockz replied:
We prefer to have these main busses as `DEPENDENCIES` so that people know what they are configuring.

e28eta committed:
Merge branch 'dev' into ms8607

e28eta committed:
Merge branch 'dev' into ms8607

e28eta committed:
Merge branch 'dev' into ms8607

e28eta committed:
Merge branch 'dev' into ms8607

e28eta committed:
Merge branch 'dev' into ms8607

e28eta committed:
Merge branch 'dev' into ms8607

e28eta committed:
Merge branch 'dev' into ms8607

e28eta committed:
Some clang-tidy fixes: making local vars const and adding virtual, public destructor

e28eta committed:
Manually applying clang-format whitespace changes

e28eta committed:
Merge branch 'dev' into ms8607

e28eta committed:
Merge branch 'dev' into ms8607

e28eta committed:
Use provided number of retries from `set_retry`

This is an update for merged PR #4251

e28eta commented:
Needs minor update for #4251 being merged

e28eta commented:
Update completed in fdbde5a, ready for review again

e28eta committed:
Merge branch 'dev' into ms8607

e28eta commented:
Still waiting on review, afaik

e28eta committed:
Merge branch 'dev' into ms8607

e28eta committed:
Merge branch 'dev' into ms8607

e28eta committed:
Merge branch 'dev' into ms8607

e28eta commented:
Hi @jesserockz,

I just wanted to make sure this hadn’t fallen into a gap. Maybe a missing label, or something else that means you wouldn’t have seen it. I assume your main priorities are bugs & features that align with the project’s strategic goals, and that makes perfect sense to me. Based on 180 open PRs, I also assume that you’ve got a really long TODO list and I don’t envy you.

 I _think_ this is in good shape and ready to merge, but if there’s anything further _I_ can do I’d love to know it.

jesserockz commented on esphome/components/ms8607/sensor.py:
```diff
@@ -0,0 +1,77 @@
+    await cg.register_component(var, config)
+    await i2c.register_i2c_device(var, config)
+
+    if CONF_TEMPERATURE in config:
+        sens = await sensor.new_sensor(config[CONF_TEMPERATURE])
```
```suggestion
    if temperature_config := config.get(CONF_TEMPERATURE):
        sens = await sensor.new_sensor(temperature_config)
```

jesserockz commented on esphome/components/ms8607/sensor.py:
```diff
@@ -0,0 +1,77 @@
+        sens = await sensor.new_sensor(config[CONF_TEMPERATURE])
+        cg.add(var.set_temperature_sensor(sens))
+
+    if CONF_PRESSURE in config:
+        sens = await sensor.new_sensor(config[CONF_PRESSURE])
```
```suggestion
    if pressure_config := config.get(CONF_PRESSURE):
        sens = await sensor.new_sensor(pressure_config)
```

jesserockz commented on esphome/components/ms8607/sensor.py:
```diff
@@ -0,0 +1,77 @@
+        sens = await sensor.new_sensor(config[CONF_PRESSURE])
+        cg.add(var.set_pressure_sensor(sens))
+
+    if CONF_HUMIDITY in config:
+        sens = await sensor.new_sensor(config[CONF_HUMIDITY])
+        cg.add(var.set_humidity_sensor(sens))
+        # C++ class creates the I2CDevice object, using I2C bus already provided to the object
+        cg.add(var.set_humidity_sensor_address(config[CONF_HUMIDITY][CONF_ADDRESS]))
```
```suggestion
    if humidity_config := config.get(CONF_HUMIDITY):
        sens = await sensor.new_sensor(humidity_config)
        cg.add(var.set_humidity_sensor(sens))
        # C++ class creates the I2CDevice object, using I2C bus already provided to the object
        cg.add(var.set_humidity_sensor_address(humidity_config[CONF_ADDRESS]))
```

e28eta committed:
Apply suggestions from code review

Co-authored-by: Jesse Hills <3060199+jesserockz@users.noreply.github.com>

e28eta committed:
Merge branch 'dev' into ms8607

e28eta committed:
Fix typo in docs on set_retry

e28eta committed:
Update set_retry call and comment based on fixes in #3305

e28eta committed:
Use separate class for Humidity sensor, created via codegen

I think this is the smallest acceptable delta to accomplish requested change from review
https://github.com/esphome/esphome/pull/3307#discussion_r1433493948

I also see an opportunity to push some logic into the newly introduced
`MS8607HumidityDevice` class, however I'm not confident that it would simplify anything.
I'm mostly happy with the flow & arrangement of logic, and I'm worried that pulling it
apart will introduce unnecessary complexity.

I'm less confident about the yaml configuration setup. This single chip has two logical
I2CDevices, and 3 Sensors, but I don't think splitting the sensors between the devices
is what a user would want to do. Instead, the Sensors live on the main component, and
the humidity I2CDevice is a mostly-hidden implementation detail.

e28eta committed:
Apply clang-format suggestions from CI

e28eta committed:
Add another test case to the yaml with more config keys used

e28eta committed:
Merge branch 'dev' into ms8607

e28eta committed:
Merge branch 'dev' into ms8607

e28eta committed:
Fix too many spaces after colon error in yaml

e28eta commented:
Thanks @jesserockz!
